### PR TITLE
Rename storage classes

### PR DESCRIPTION
**Pull request by erikjohnston**
As per #11733

### CODE REVIEW

**Review by richvdh — CHANGES_REQUESTED** [review]
https://github.com/matrix-org/synapse/issues/11733#issuecomment-1138828890 suggests some changes to the packages too. Please can they happen?

**Comment by erikjohnston** [comment]
Yay, this is going to cause all the merge conflicts.

File: synapse/federation/federation_server.py [hunk 1]
```diff
@@ -109,7 +109,7 @@ def __init__(self, hs: "HomeServer"):
         super().__init__(hs)
 
         self.handler = hs.get_federation_handler()
-        self.storage = hs.get_storage()
+        self.storage_controllers = hs.get_storage_controllers()
```
Comment (richvdh) [inline synapse/federation/federation_server.py:112]: if this isn't used, let's get rid of it.

**Review by richvdh — COMMENTED** [review]
while you're renaming properties, please can you give them underscore prefixes?

File: synapse/state/__init__.py [hunk 2]
```diff
@@ -127,10 +127,10 @@ class StateHandler:
     def __init__(self, hs: "HomeServer"):
         self.clock = hs.get_clock()
         self.store = hs.get_datastores().main
-        self.state_storage = hs.get_storage().state
+        self.state_storage_controller = hs.get_storage_controllers().state
```
Comment (richvdh) [inline synapse/state/__init__.py:130]:
```suggestion
        self._state_storage_controller = hs.get_storage_controllers().state
```

and updates below

File: synapse/storage/controllers/state.py [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+    Tuple,
+)
+
+from synapse.events import EventBase
+from synapse.storage.state import StateFilter
+from synapse.storage.util.partial_state_events_tracker import PartialStateEventsTracker
+from synapse.types import MutableStateMap, StateMap
+
+if TYPE_CHECKING:
+
```
Comment (richvdh) [inline synapse/storage/controllers/state.py:33]:
```suggestion
```

File: synapse/storage/state.py [hunk 4]
```diff
@@ -578,318 +573,3 @@ def must_await_full_state(self, is_mine_id: Callable[[str], bool]) -> bool:
     types=frozendict({EventTypes.Member: frozenset()}), include_others=True
 )
 _NONE_STATE_FILTER = StateFilter(types=frozendict(), include_others=False)
```
Comment (richvdh) [inline synapse/storage/state.py:575]: I wonder if we ought to move the `StateFilter` stuff too. Is it used anywhere other than to pass down to `StateGroupStorage` ?

**Review by richvdh — COMMENTED** [review]
lgtm in general.

Please update the docstring at https://github.com/matrix-org/synapse/blob/4302238a744ac5dd5492f36eab1efa3a83ec9d8e/synapse/storage/__init__.py#L17 to mention storage controllers.

File: synapse/storage/state.py (hunk 4 above)
Comment (erikjohnston) [reply synapse/storage/state.py:575]: I'd be tempted to leave it where it is TBH, we tend to define types that are external to the `synapse.storage` module in top level files :shrug:

Comment (richvdh) [reply synapse/storage/state.py:575]: I don't disagree, but I don't follow your reasoning.

> we tend to define types that are external to the synapse.storage module in top level files

... like what? And this is a type internal to the `synapse.storage` module, so how is it relevant?

Comment (erikjohnston) [reply synapse/storage/state.py:575]: `synapse.storage.roommember` has a bunch of types in it for example?



> And this is a type internal to the `synapse.storage` module, so how is it relevant?

The `StateFilter` is used all over the code base to specify the filter. If you mean `_NONE_STATE_FILTER` and co, then they are used solely by the `StateFilter` class?

File: synapse/storage/__init__.py [hunk 5]
```diff
@@ -20,39 +20,16 @@
 database. The `databases` are classes that talk directly to a `DatabasePool`
 instance and have associated schemas, background updates, etc. On top of those
 there are classes that provide high level interfaces that combine calls to
-multiple `databases`.
+multiple `databases`, called storage controllers and are located in the
+`controller` module. These are bundled into a single `StorageControllers` class
+for ease of use and exposed via `HomeServer.get_storage_controllers()`.
```
Comment (richvdh) [inline synapse/storage/__init__.py:25]:
```suggestion
instance and have associated schemas, background updates, etc. 

On top of the databases are the StorageControllers, located in the
`synapse.storage.controllers` module. These classes provide high level interfaces
that combine calls to multiple `databases`. They are bundled into the
`StorageControllers` singleton for ease of use, and exposed via
`HomeServer.get_storage_controllers()`.
```

**Review by richvdh — APPROVED** [review]
some word-wrangling, lgtm otherwise

File: synapse/storage/state.py (hunk 4 above)
Comment (richvdh) [reply synapse/storage/state.py:575]: oh right. by "external" you mean "part of the external api", sorry.

I'm not convinced that's a convention we actually follow, but I don't mind it staying here for now anyway.

**Pull request merged**